packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -306,14 +319,89 @@ export class PHPRequestHandler {
 			),
 			this.rewriteRules
 		);
-		const fsPath = joinPaths(this.#DOCROOT, normalizedRequestedPath);
-		if (!seemsLikeAPHPRequestHandlerPath(fsPath)) {
-			return this.#serveStaticFile(
-				await this.processManager.getPrimaryPhp(),
-				fsPath
-			);
+
+		const primaryPhp = await this.getPrimaryPhp();
+
+		let fsPath = joinPaths(this.#DOCROOT, normalizedRequestedPath);
+
+		if (primaryPhp.isDir(fsPath)) {
+			// Ensure directory URIs have a trailing slash. Otherwise,
+			// relative URIs in index.php or index.html files are relative
+			// to the next directory up.
+			//
+			// Example:
+			// For a request to "/wp-admin", the relative link "edit.php"
+			// resolves to "/edit.php" rather than "/wp-admin/edit.php".
+			//
+			// This is correct behavior for the browser:
+			// https://www.rfc-editor.org/rfc/rfc3986#section-5.2.3
+			//
+			// But the intent for `/wp-admin/index.php` is that its relative
+			// URIs are relative to `/wp-admin/`.
+			//
+			// In fact, WordPress also redirects like this when given a chance.
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L696
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L1036-L1045
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/link-template.php#L3558
+			if (!fsPath.endsWith('/')) {
+				return new PHPResponse(
+					301,
+					{ Location: [`${requestedUrl.pathname}/`] },
+					new Uint8Array(0)
+				);
+			}
+
+			// We can only satisfy requests for directories with a default file
+			// so let's first resolve to a default path when available.
+			const defaultFilePath = ['index.php', 'index.html']
```

Comment:
Eventually we may want to make it configurable, but I think we're good for now.

Reply:
Code style nit – `defaultFilePath` doesn't suggest we're testing whether the file exists or not. How about changing this to a loop?

```ts
// We can only satisfy requests for directories with a default file
// so let's first resolve to a default path when available.
for ( const possibleIndexFile of ['index.php', 'index.html'] ) {
    const possibleIndexPath = joinPaths(fsPath, possibleIndexFile);
    if ( primaryPhp.isFile( possibleIndexPath ) ) {
        fsPath = possibleIndexPath;
        break;
    }
}
```

Reply:
> Code style nit – defaultFilePath doesn't suggest we're testing whether the file exists or not. How about changing this to a loop to avoid that naming conundrum entirely?

Updated!